src/sparse_blas/backends/rocsparse/operations/rocsparse_spmv.cpp

```diff
@@ -0,0 +1,350 @@
+                          oneapi::mkl::sparse::spmv_descr_t spmv_descr, void* workspace,
+                          const std::vector<sycl::event>& dependencies) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    if (A_handle->all_use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    detail::common_spmv_optimize(opA, is_alpha_host_accessible, A_view, A_handle, x_handle,
+                                 is_beta_host_accessible, y_handle, alg, spmv_descr);
+    spmv_descr->workspace.usm_ptr = workspace;
+    if (alg == oneapi::mkl::sparse::spmv_alg::no_optimize_alg) {
+        return detail::collapse_dependencies(queue, dependencies);
+    }
+    std::size_t buffer_size = spmv_descr->temp_buffer_size;
+    auto functor = [=](sycl::interop_handle) {
+        auto roc_handle = spmv_descr->roc_handle;
+        detail::spmv_optimize_impl(roc_handle, opA, alpha, A_handle, x_handle, beta, y_handle, alg,
+                                   buffer_size, workspace, is_alpha_host_accessible);
+    };
+
+    return detail::dispatch_submit(__func__, queue, dependencies, functor, A_handle, x_handle,
+                                   y_handle);
+}
+
+sycl::event spmv(sycl::queue& queue, oneapi::mkl::transpose opA, const void* alpha,
+                 oneapi::mkl::sparse::matrix_view A_view,
+                 oneapi::mkl::sparse::matrix_handle_t A_handle,
+                 oneapi::mkl::sparse::dense_vector_handle_t x_handle, const void* beta,
+                 oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                 oneapi::mkl::sparse::spmv_alg alg, oneapi::mkl::sparse::spmv_descr_t spmv_descr,
+                 const std::vector<sycl::event>& dependencies) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    if (A_handle->all_use_buffer() != spmv_descr->workspace.use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    detail::check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle,
+                             is_alpha_host_accessible, is_beta_host_accessible);
+
+    if (!spmv_descr->optimized_called) {
+        throw mkl::uninitialized(
+            "sparse_blas", __func__,
+            "spmv_optimize must be called with the same arguments before spmv.");
+    }
+    CHECK_DESCR_MATCH(spmv_descr, opA, "spmv_optimize");
+    CHECK_DESCR_MATCH(spmv_descr, A_view, "spmv_optimize");
+    CHECK_DESCR_MATCH(spmv_descr, A_handle, "spmv_optimize");
+    CHECK_DESCR_MATCH(spmv_descr, x_handle, "spmv_optimize");
+    CHECK_DESCR_MATCH(spmv_descr, y_handle, "spmv_optimize");
+    CHECK_DESCR_MATCH(spmv_descr, alg, "spmv_optimize");
+
+    A_handle->mark_used();
+    auto& buffer_size = spmv_descr->temp_buffer_size;
```

Comment:
Is there a reason why this variable specifically is kept and captured by reference _outside_ `compute_functor` while other members of `spmv_descr` are created inside the functor? If it is because `buffer_size` is used later in an `if` condition, I'd prefer it if we change this approach and capture everything by `[=]` in `compute_functor`, even though it would mean replacing the last `buffer_size > 0` with `spmv_descr->temp_buffer_size > 0`. I know that what you have right now is expected to work in this particular case (because of the condition that `spmv_descr` must live as long as `spmv` is being called), it is normally a bad idea to capture variables by reference in SYCL functors that are going to be running asynchronously unless you have an immediate `event.wait()` (e.g., like what we are doing in `spmv_buffer_size()` function).

Reply:
Sure that's fine, done in https://github.com/oneapi-src/oneMKL/pull/544/commits/c53ef8d2cc49f9ec695b1d01bd49d238513ac42d